docs/rfc/0002-plugins.md

```diff
@@ -0,0 +1,184 @@
+- The plugin process returns an answer
+- The answer is serialized and sent back to the `js` vm
+
+### The standard library `plugin` function
+
+The core `plugin` RPC call is quite general:
+
+```text
+plugin(kind: string, url: string, input: JSON) -> JSON
+```
+
+- `kind` is the kind of plugin invoked (eg. `render`). Plugin binaries can
+implement more than one kind of plugins.
+- `url` identifies a plugin, for instance:
+`https://jkcfg.github.io/plugins/helm/0.1.0/plugin.json`. This JSON file is
+really plugin metadata (see next section). It is highly recommended, for
+reproducibility, to ensure the plugin definition and binaries the URL points
+to be immutable to encode a version in the URL.
+- `input` and the return value are generic JSON objects that wrapping library
+code are responsible for understanding.
+
+### Plugin definition
+
+The plugin URL in the `plugin` call points to a JSON file describing the plugin:
+
+```json
+{
+  "name": "helm",
+  "version": "0.1.0",
+  "binaries": {
+    "linux-amd64": "https://jkcfg.github.io/plugins/helm/0.1.0/linux-amd64/jk-plugin-helm",
+    "darwin-amd64": "https://jkcfg.github.io/plugins/helm/0.1.0/darwin-amd64/jk-plugin-helm"
+  }
+}
+```
+
+### Local plugins
+
+For development purposes it is possible to point the `plugin` RPC call to a
+local file by giving a relative path to the JSON plugin definition. The
+`binaries` fields can point at plugins present in the `PATH`.
+
+```json
+{
+  "name": "echo",
+  "version": "0.1.0",
+  "binaries": {
+    "linux-amd64": "jk-plugin-echo",
+    "darwin-amd64": "jk-plugin-echo"
+  }
+}
+```
+
+## Backward compatibility
+
+New feature, no backward compatibility concerns.
+
+## Drawbacks and limitations
+
+### Complexity
+
+Plugins do add more complexity to `jk`. Complexity creep is somewhat
+unavoidable if we want to support things such as consuming helm charts. One
+good thing about plugins is that, at the cost of a (simple) interface between
+`jk` and plugins, most of the complexity is delegated to plugin code, not the
+core.
+
+We should ensure that plugins don't add any cognitive load on the user. By
+wrapping plugin invocation in library objects we can make then mostly
+transparent to the user with, maybe, the exception of plugin downloading.
+
+## Hermeticity
+
+Plugins have a high potential to break hermeticity. We should ensure our
+plugins are made "in good faith", are self-contained and as deterministic as
+possible.
+
+I believe that's an ok price to pay for such extensibility power.
+
+## Alternatives
+
+- A generic `exec` standard library function that would execute any binary in
+  the path.
+
+  Problems with that approach:
+
+  1. Executing whatever is in the path doesn't play well with hermeticity.
+  Plugins have versions for reproducibility.
+  1. Packaging. One goal is to be able to package all the dependencies needed
+  for a `jk` script to run. Having the plugin abstraction with metadata
+  allows that.
+
```

Comment:
I reckon it's worth mentioning `|` here -- i.e., no plugins, just pipe the output of another command into `jk` (or write files and use them as input to `jk`, etc.). One of the problems I remember from our discussions of this is that you can't have more than one source of input via `|`; another (slightly harder to pin down) is that it's better to have control from within JavaScript code, so you can e.g., encapsulate the fact of the external tooling.